ocaml/typing/jkind_axis.ml

```diff
@@ -0,0 +1,225 @@
+    match t1, t2 with
+    | External, External -> Equal
+    | External, (External64 | Internal) -> Less
+    | External64, External -> Not_le
+    | External64, External64 -> Equal
+    | External64, Internal -> Less
+    | Internal, (External | External64) -> Not_le
+    | Internal, Internal -> Equal
+
+  let le t1 t2 = Misc.Le_result.is_le (less_or_equal t1 t2)
+
+  let meet t1 t2 =
+    match t1, t2 with
+    | External, (External | External64 | Internal)
+    | (External64 | Internal), External ->
+      External
+    | External64, (External64 | Internal) | Internal, External64 -> External64
+    | Internal, Internal -> Internal
+
+  let print ppf = function
+    | External -> Format.fprintf ppf "external_"
+    | External64 -> Format.fprintf ppf "external64"
+    | Internal -> Format.fprintf ppf "internal"
+end
+
+module Nullability = struct
+  type t =
+    | Non_null
+    | Maybe_null
+
+  let max = Maybe_null
+
+  let min = Non_null
+
+  let equal n1 n2 =
+    match n1, n2 with
+    | Non_null, Non_null -> true
+    | Maybe_null, Maybe_null -> true
+    | (Non_null | Maybe_null), _ -> false
+
+  let less_or_equal n1 n2 : Misc.Le_result.t =
+    match n1, n2 with
+    | Non_null, Non_null -> Equal
+    | Non_null, Maybe_null -> Less
+    | Maybe_null, Non_null -> Not_le
+    | Maybe_null, Maybe_null -> Equal
+
+  let le n1 n2 = Misc.Le_result.is_le (less_or_equal n1 n2)
+
+  let meet n1 n2 =
+    match n1, n2 with
+    | Non_null, (Non_null | Maybe_null) | Maybe_null, Non_null -> Non_null
+    | Maybe_null, Maybe_null -> Maybe_null
+
+  let print ppf = function
+    | Non_null -> Format.fprintf ppf "non_null"
+    | Maybe_null -> Format.fprintf ppf "maybe_null"
+end
+
+module type Axis_s = sig
+  type t
+
+  val max : t
+
+  val min : t
+
+  val equal : t -> t -> bool
+
+  val less_or_equal : t -> t -> Misc.Le_result.t
+
+  val le : t -> t -> bool
+
+  val meet : t -> t -> t
+
+  val print : Format.formatter -> t -> unit
+end
+
+module Axis = struct
+  module Modal = struct
+    type 'a t =
+      | Locality : Mode.Locality.Const.t t
+      | Linearity : Mode.Linearity.Const.t t
+      | Uniqueness : Mode.Uniqueness.Const.t t
+      | Portability : Mode.Portability.Const.t t
+      | Contention : Mode.Contention.Const.t t
+  end
+
+  module Nonmodal = struct
+    type 'a t =
+      | Externality : Externality.t t
+      | Nullability : Nullability.t t
+  end
+
+  type 'a t =
+    | Modal of 'a Modal.t
+    | Nonmodal of 'a Nonmodal.t
+
+  type packed = Pack : 'a t -> packed
+
+  module Accent_lattice (M : Mode_intf.Lattice) = struct
+    (* A functor to add some convenient functions to modal axes *)
+    include M
+
+    let less_or_equal a b : Misc.Le_result.t =
+      match le a b, le b a with
+      | true, true -> Equal
+      | true, false -> Less
+      | false, _ -> Not_le
+
+    let equal a b = Misc.Le_result.is_equal (less_or_equal a b)
+  end
+
+  let get (type a) : a t -> (module Axis_s with type t = a) = function
```

Comment:
I'm slightly concerned about the first-class module here. It's later used by `with` syntax PR in `Bounds.sub`, which is hot-path? This could be costy.

One solution is to push `Axis.t` inside the module. That is, having this in a module:
```
val less_or_equal : t Axis.t -> t -> t -> Misc.Le_result.t
```